staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go

```diff
@@ -0,0 +1,336 @@
+	}, r.Options...)
+
+	s := validate.NewSchemaValidator(r.Schema, r.Root, r.Path, r.KnownFormats, opts...)
+
+	res := s.Validate(value)
+
+	if res.IsValid() {
+		return res
+	}
+
+	// Current policy is to ratchet errors if the comparison is a noop
+	if r.currentInvocation.IsNoopLazy() {
+		newRes := &validate.Result{}
+		newRes.MergeAsWarnings(res)
+		return newRes
+	}
+
+	return res
+}
+
+func (r *ratchetingValueValidator) SubPropertyValidator(value interface{}, field string, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, options ...validate.Option) validate.ValueValidator {
+	// Find correlated old value
+	if asMap, ok := r.oldValue.(map[string]interface{}); ok {
+		return newRatchetingValueValidator(
+			asMap[field],
+			schemaArgs{
+				Schema:       schema,
+				Root:         rootSchema,
+				Path:         root,
+				KnownFormats: formats,
+				Options:      options,
+			},
+			r,
+			field,
+		)
+	}
+
+	return validate.NewSchemaValidator(schema, rootSchema, root, formats, options...)
+
+}
+
+func (r *ratchetingValueValidator) SubIndexValidator(value interface{}, index int, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, options ...validate.Option) validate.ValueValidator {
+	oldAsList, ok := r.oldValue.([]interface{})
+	if !ok {
+		return validate.NewSchemaValidator(schema, rootSchema, root, formats, options...)
+	}
+
+	asList, ok := value.([]interface{})
+	if !ok {
+		return validate.NewSchemaValidator(schema, rootSchema, root, formats, options...)
+	} else if len(asList) <= index {
+		// Skip ratcheting if for some reason the index if invalid (shouldn't happen)
+		return validate.NewSchemaValidator(schema, rootSchema, root, formats, options...)
+	}
+
+	listType, _ := schema.Extensions.GetString("x-kubernetes-list-type")
+	switch listType {
+	case "map":
+		// Look up keys for this index in current object
+		currentElement := asList[index]
+
+		oldList := r.currentInvocation.mapList
+		if oldList == nil {
+			oldList = celopenapi.MakeMapList(schema, oldAsList)
+			r.currentInvocation.mapList = oldList
+		}
+		return newRatchetingValueValidator(
+			oldList.Get(currentElement),
+			schemaArgs{
+				Schema:       schema,
+				Root:         rootSchema,
+				Path:         root,
+				KnownFormats: formats,
+				Options:      options,
+			},
+			r,
+			index,
+		)
+
+	case "set":
+		// Are sets correlatable? Only if the old value equals the current value.
+		// We might be able to support this, but do not currently see a lot
+		// of value
+		// (would allow you to add/remove items from sets with ratcheting but not change them)
+		return validate.NewSchemaValidator(schema, rootSchema, root, formats, options...)
+	case "atomic":
+		// Atomic lists are not correlatable by item
+		// Ratcheting is not available on a per-index basis
+		return validate.NewSchemaValidator(schema, rootSchema, root, formats, options...)
+	default:
+		if len(oldAsList) > index {
```

Comment:
nit: switching around those two return would easier to read.

Reply:
done